internal/transport/transport_test.go

```diff
@@ -2746,3 +2746,44 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 		t.Errorf("Context timed out")
 	}
 }
+
+// Tests the scenario where a client transport is closed and writing of the
+// GOAWAY frame as part of the close does not complete because of a network
+// hang. The test verifies that the client transport is closed without waiting
+// for too long.
+func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
+	// Override timer for writing GOAWAY to 0 so that the connection write
+	// always times out. It is equivalent of real network hang when conn
+	// write for goaway doesn't finish in specified deadline
```

Comment:
This feels like a race?  How can we be sure that the loopyWriter doesn't process the item to exit before the timer is even created?  We need to make the Conn hang on demand, not just override the timeout to zero.  I think we can do this pretty easily by wrapping the conn by injecting a dialer, and set a flag for when we want `Write` to just not do anything.

Reply:
Umm.... I initially had the custom dialer with a hang [here](https://github.com/grpc/grpc-go/blob/59604784b6641bf7df8c3f270f8574915bae54ed/internal/transport/transport_test.go#L2734) but that will require the test to run for few seconds which may cause flakiness. Hence, I decided to override the timeout to 0 so that write is always expected to be time out. WDYT?

Reply:
I'm fine with an instant (or short?) timeout, but this doesn't actually make the connection hang, so it doesn't actually test anything, does it?  I.e. this test will pass even before your changes, right?  (Obviously with the exception that it sets a global variable that doesn't exist @ HEAD.)

Reply:
Ah makes sense.  I have updated the logic of custom dialer now, which hangs the connection.